File: gunkim/today_problem/23_11/1189_컴백홈.java
// 1 try 1 solve
// dfs 정석 문제이다.
import java.util.*;
import java.io.*;

public class Main {

    private int answer = 0;
    private int r;
    private int c;
    private int k;
    private boolean[][] visit;
    private char[][] board;

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        board = new char[r][c];
        for (int i = 0; i < r; i++) {
            String input = br.readLine();
            for (int j = 0; j < c; j++) {
                board[i][j] = input.charAt(j);
            }
        }

        visit = new boolean[r][c];
        visit[r - 1][0] = true;
        dfs(r - 1, 0 ,1);


        System.out.println(answer);
    }


    private int[][] dir = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public void dfs(int y, int x, int count) {
        // dfs 탈출 조건
        if (count == k) {
            if (y == 0 && x == c - 1) answer++;
            return;
        }

        for (int i = 0; i < 4; i++) {
            int ny = y + dir[i][0];
            int nx = x + dir[i][1];

            if(ny < 0 || ny >= r || nx < 0 || nx >= c || visit[ny][nx]) continue;
            if(board[ny][nx] == 'T') continue; // 벽으로 이동할 수는 없음

            visit[ny][nx] = true;
            dfs(ny, nx, count + 1);
            visit[ny][nx] = false;
        }
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

}

File: LeeYooseok/BOJ_오늘의 문제/2023/2023-10-26/1189 컴백홈.py
# DFS
R, C, K = map(int, input().split())

board = [list(input()) for _ in range(R)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

answer = 0
visited = [[False for _ in range(C)] for _ in range(R)]

def dfs(x, y, distance):
    global answer

    if distance == K and x == C - 1 and y == 0:
        answer += 1
    else:
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            # 백트래킹 한정 조건 : board[nx][ny] == '.'
            if 0 <= nx < C and 0 <= ny < R and board[ny][nx] == '.' and not visited[ny][nx]:
                visited[ny][nx] = True
                dfs(nx, ny, distance + 1)
                # 원래 상태로 돌려 놓아 다른 방향으로 다시 탐색하도록 한다.
                visited[ny][nx] = False


visited[R - 1][0] = True
dfs(0, R - 1, 1)
# 정답
print(answer)
